cosmwasm/enclaves/shared/contract-engine/src/contract_validation.rs

```diff
@@ -616,53 +810,335 @@ fn verify_contract(msg: &CosmWasmMsg, contract_address: &HumanAddr) -> bool {
+                    }
+                };
+
+                // memo must be set in ibc-hooks
+                let memo = match packet_data.memo {
+                    Some(memo) => memo,
+                    None => {
+                        trace!("Contract was called via ibc-hooks ack callback but packet_data.memo is empty");
+                        return false;
+                    }
+                };
+
+                // Parse data.memo as `{"ibc_callback": "secret1contractAddr"}` JSON
+                let ibc_hooks_outgoing_memo: IbcHooksOutgoingTransferMemo =
+                    match serde_json::from_slice(memo.as_bytes()) {
+                        Ok(wasm_msg) => wasm_msg,
+                        Err(err) => {
+                            trace!(
+                                "Contract was called via ibc-hooks but packet_data.memo cannot be parsed as IbcHooksWasmMsg: {:?} Error: {:?}",
+                                memo,
+                                err,
+                            );
+                            return false;
+                        }
+                    };
+
+                let is_verified = *contract_address == ibc_hooks_outgoing_memo.ibc_callback
+                    && *contract_address == packet_data.sender;
+                if !is_verified {
+                    trace!(
+                        "Contract address sent to enclave {:?} is not the same as in ibc-hooks outgoing transfer callback address packet {:?}",
+                        contract_address,
+                        ibc_hooks_outgoing_memo.ibc_callback
+                    );
+                }
+                is_verified
+            } else {
+                // Packet was sent from an IBC enabled contract
+                // source_port is of the form "wasm.{contract_address}"
+
+                // Extract contract_address from source_port
+                // This also checks that source_port starts with "wasm."
+                let contract_address_from_port = match source_port.strip_prefix("wasm.") {
+                    Some(contract_address) => contract_address,
+                    None => {
+                        trace!(
+                            "IBC-enabled Contract was called via MsgAcknowledgement but source_port doesn't start with \"wasm.\": {:?}",
+                            source_port,
+                        );
+                        return false;
+                    }
+                };
+
+                let is_verified = *contract_address == HumanAddr::from(contract_address_from_port);
+                if !is_verified {
+                    trace!(
+                        "Contract address sent to enclave {:?} is not the same as extracted from MsgAcknowledgement but source_port: {:?}",
+                        contract_address,
+                        contract_address_from_port,
+                    );
+                }
+                is_verified
+            }
+        }
```

Comment:
Needs more functions...

Reply:
Same goes for the rest of the new verification code (verify_sent_funds) etc

Reply:
https://github.com/scrtlabs/SecretNetwork/commit/c3addbc8b